docs/proposals/20221030-garbage-collection-of-tenant-resources.md

```diff
@@ -0,0 +1,153 @@
+
+- When an `UPDATE` event is watched, if the Tenant has a `DeletionTimestamp` and a `kubezoo.io/tenant` finalizer,
+  
+    1. List all cluster-scoped GVRs via DiscoveryClient, and divide them into two groups according to whether they are CRD or not.
+       
+    2. Traverse the GVR list of CRDs. For each GVR, check whether the CRD belongs to this tenant according to the prefix of its `Group`. If so, delete it via DynamicClient.
+
+        > In this step, CRs associated with CRDs belonging to this tenant will be deleted in cascade. This avoids the overhead of listing all CRs in step (iii) and improves performance.
+    
+    3. Traverse the list of non-CRD GVRs. For each GVR, list all objects of this kind via DynamicClient. For each object, check whether it belongs to this tenant according to the prefix of its `Name`. If so, delete it via DynamicClient.
+        > - In this step, the tenant's namespaces will be cleaned up, which triggers the Namespace Controller to clean up the tenant's Namespaced resources.
+        > - In this step, the tenant's CRs associated with system CRDs are also cleaned up.
+
+    4. Remove the `kubezoo.io/tenant` finalizer from the Tenant object.
+    
+## Alternatives
+
+- **Delete tenant's cluster-scoped resources in cascade through the OwnerReferences mechanism**
+
+    - When a `CREATE` event of a Tenant object is watched, Tenant Controller will
+        - Create a dummy cluster-scoped object for the tenant in the upstream cluster, and record the `UID` of this dummy object in the Tenant object.
+    
+        - Add a `kubezoo.io/tenant` finalizer for the Tenant object.
+
+    - When a tenant creates a cluster-scoped resource through KubeZoo, the proxy adds an `OwnerReference` (pointing to the dummy object) to it before creating it in the upstream cluster.
+    
+    - When a tenant queries cluster-scoped resources through KubeZoo, the proxy removes the `OwnerReference` added by KubeZoo and returns it to the tenant.
+
+    - When a `DELETE` event of a Tenant object is watched, if the Tenant has a `DeletionTimestamp` and a `kubezoo.io/tenant` finalizer, Tenant Controller will
+        
+        - Delete the dummy object in the upstream cluster, which triggers the Garbage Collector to clean up the tenant's cluster-scoped resources.
+    
+        - Remove the `kubezoo.io/tenant` finalizer from the Tenant object. 
+          
+        This alternative has some flaws:
+      
+        - More intrusive: We need to modify the Tenant Controller, protocol conversion rules, and the definition of Tenant object. Besides, we also need to create additional objects in the upstream cluster.
+    
+        - High risk of accidental deletion: All cluster-scoped resources of a tenant are associated with a single dummy object. If the dummy object is deleted by mistake, all the resources of the tenant will be deleted in cascade, which has a large impact.
```

Comment:
Doesn't this apply to the case where we delete a Tenant object by mistake as well?

Reply:
Yes, it applies as well. Introducing a new object in the upstream cluster may increase the risk.